File: pirellone/Java/skeleton.java
import java.util.Scanner;

abstract class Skeleton {
    private static final Scanner in = new Scanner(System.in);


    abstract int is_solvable(int M, int N, int P[][]);

    interface SolveCallbacks {
        void switch_row(int i);
        void switch_col(int j);
    }
    abstract void solve(int M, int N, int P[][], SolveCallbacks callbacks);

    public static void main(String[] args) {
        Solution __solution = new Solution();

        // checkpoint
        System.out.printf("%d\n", 0);
        // read choice
        int choice;
        System.out.flush();
        choice = in.nextInt();
        // if choice {...} else {...}
        if (choice != 0) {
            // read M, N
            int M;
            int N;
            System.out.flush();
            M = in.nextInt();
            N = in.nextInt();
            // for i to M {...}
            int[][] P;
            P = new int[M][];
            for (int i = 0; i < M; i++) {
                // for j to N {...}
                P[i] = new int[N];
                for (int j = 0; j < N; j++) {
                    // read P[i][j]
                    System.out.flush();
                    P[i][j] = in.nextInt();
                }
            }
            // call res = is_solvable(M, N, P)
            int res;
            res = __solution.is_solvable(M, N, P);
            // write res
            System.out.printf("%d\n", res);
        } else {
            // read M, N
            int M;
            int N;
            System.out.flush();
            M = in.nextInt();
            N = in.nextInt();
            // for i to M {...}
            int[][] P;
            P = new int[M][];
            for (int i = 0; i < M; i++) {
                // for j to N {...}
                P[i] = new int[N];
                for (int j = 0; j < N; j++) {
                    // read P[i][j]
                    System.out.flush();
                    P[i][j] = in.nextInt();
                }
            }
            // call solve(M, N, P) callbacks {...}
            SolveCallbacks solveCallbacks = new SolveCallbacks() {
                public void switch_row(int i) {
                    // callback switch_row
                    System.out.printf("%d %d\n", 1, 0);
                    // write i
                    System.out.printf("%d\n", i);
                }
                public void switch_col(int j) {
                    // callback switch_col
                    System.out.printf("%d %d\n", 1, 1);
                    // write j
                    System.out.printf("%d\n", j);
                }
            };
            __solution.solve(M, N, P,  solveCallbacks);
            // no more callbacks
            System.out.printf("%d %d\n", 0, 0);
        }
        // exit
        System.exit(0);
    }
}

File: pirellone/Python/skeleton.py
import os as _os

def main(_solution):
    # checkpoint
    print(0)
    # read choice
    print(end="", flush=True)
    [choice] = map(int, input().split())
    # if choice {...} else {...}
    if choice:
        # read M, N
        print(end="", flush=True)
        [M, N] = map(int, input().split())
        # for i to M {...}
        P = [None] * M
        for i in range(M):
            # for j to N {...}
            P[i] = [None] * N
            for j in range(N):
                # read P[i][j]
                print(end="", flush=True)
                [P[i][j]] = map(int, input().split())
        # call res = is_solvable(M, N, P)
        res = int(_solution.is_solvable(M, N, P))
        # write res
        print(res)
    else:
        # read M, N
        print(end="", flush=True)
        [M, N] = map(int, input().split())
        # for i to M {...}
        P = [None] * M
        for i in range(M):
            # for j to N {...}
            P[i] = [None] * N
            for j in range(N):
                # read P[i][j]
                print(end="", flush=True)
                [P[i][j]] = map(int, input().split())
        # call solve(M, N, P) callbacks {...}
        def _callback_switch_row(i):
            # callback switch_row
            print(1, 0)
            # write i
            print(i)
        def _callback_switch_col(j):
            # callback switch_col
            print(1, 1)
            # write j
            print(j)
        _solution.solve(M, N, P, _callback_switch_row, _callback_switch_col)
        # no more callbacks
        print(0, 0)
    # exit
    print(end="", flush=True)
    _os._exit(0)


if __name__ == '__main__':
    import sys
    import runpy
    
    if len(sys.argv) != 2:
        print("Usage: {} <solution>".format(sys.argv[0]))
    
    class Wrapper: pass 
    solution = Wrapper()
    solution.__dict__ = runpy.run_path(sys.argv[1])
    main(solution)

